### ISSUE TITLE
Can I incrementally update tiles?

### ISSUE THREAD

**Issue by eamonnmcevoy** [valhalla/valhalla#3386]
I would like to incrementally rebuild subsets of routing tiles.

Is this scenario possible?

1. build a full set of tiles covering the planet
```
 valhalla_build_tiles -c ./conf/valhalla.json planet.osm.pbf
```

2. receive a subset of changes for a region and rebuild
```
 valhalla_build_tiles -c ./conf/valhalla.json planet.osm.pbf z_updated_berlin_road_network.osm.pbf
```

In this case I believe the `valhalla_build_tiles` tool will regenerate all tiles. Is there any way to have it skip tiles that are already generated and unchanged?

**Comment by kevinkreiser** [comment]
@eamonnmcevoy sadly no. that is currently something that doesnt work the way you would expect.

with the second commaand you posted what will happen is it will build both the planet and the updated part of berlin again. where ways are duplicated they will exist twice. if connectivity has changed in some way between the two both versions will exist in the final data set. and most annoyingly, it will take the entire runtime that the first run took.

when we came up with this scheme we really hoped that incremental would be possible. and it totally is, we should be able to stitch tilesets together given enough metadata between tilesets. in this way we could even do distributed planet builds and build the planet much faster. but before we get there we need someone to spend time on writing a stitcher AND the stitcher has to be fast or it wont be worth it. once uppon a time i did a little write up about the idea but i never posted it on github. let me add a placeholder issue to talk about that. or perhaps ill reformat this issue to ask for this enhancement.

**Comment by eamonnmcevoy** [comment]
Thanks for the info @kevinkreiser you saved me a lot of time experimenting with that approach.

If, for example, the z_updated_berlin_road_network.osm.pbf file had the dimensions of one Valhalla tile exactly. Could I generate one tile and replace the old one? Or, as I suspect, is the tile generation affected by the presence of neighboring tiles?

Hopefully that question make sense.

**Comment by kevinkreiser** [comment]
@eamonnmcevoy you could do that but they wont connect to each other properly. the connectivity between the adjacent tiles might point to the wrong spots in the new tile. and so you'd get crazy routes most likely

**Comment by eamonnmcevoy** [comment]
That makes sense, we should need to ensure that the edges of the tile haven't changed.

**Comment by kevinkreiser** [comment]
@eamonnmcevoy the ones that cross the tile boundary yes. really what needs to be done is they need to be matched and then they need to be stitched together (moved to the right place in the tiles data arrangement) so that the tiles are "compatible" with each other again

**Comment by swapnilmarotidukare-tomtom** [comment]
> @eamonnmcevoy sadly no. that is currently something that doesnt work the way you would expect.
> 
> with the second commaand you posted what will happen is it will build both the planet and the updated part of berlin again. where ways are duplicated they will exist twice. if connectivity has changed in some way between the two both versions will exist in the final data set. and most annoyingly, it will take the entire runtime that the first run took.
> 
> when we came up with this scheme we really hoped that incremental would be possible. and it totally is, we should be able to stitch tilesets together given enough metadata between tilesets. in this way we could even do distributed planet builds and build the planet much faster. but before we get there we need someone to spend time on writing a stitcher AND the stitcher has to be fast or it wont be worth it. once uppon a time i did a little write up about the idea but i never posted it on github. let me add a placeholder issue to talk about that. or perhaps ill reformat this issue to ask for this enhancement.

Hi @kevinkreiser any update on the stitcher, or any other way you would suggest to build the planet faster?

**Comment by nilsnolde** [comment]
Incremental builds would be awesome but won’t happen for a longer time unless someone steps up to do or fund it. 

So for now you only get a bigger machine

**Comment by kevinkreiser** [comment]
yeah sorry no progress here. it would definitely be an interesting research project to see about doing it though as its not only useful for the server side but particularly useful for the client side. if anyone is interested in doing some experiments on it let me know and i can get you started. until then we'll have to wait for someone to get some free time as @nilsnolde  mentioned.
